### PR TITLE
Auto-fix for vue/component-tags-order

### LINKED ISSUE #1463: vue/component-tags-order - Could it be made fixable?
**What rule do you want to change?**

`vue/component-tags-order`

**Does this change cause the rule to produce more or fewer warnings?**

No

**How will the change be implemented? (New option, new default behavior, etc.)?**

I wish this rule could be auto-changed on `--fix`

**Please provide some example code that this change will affect:**

None. Rule stays the same.

**What does the rule currently do for this code?**

No changes.

**What will the rule do after it's changed?**

On --fix, it will swap a single-file component's main tags around so they're in the order prescribed by the rule.

**Additional context**

I've been thinking of putting together a regex that would allow me to run a mass replacement on my project to ensure all tags are in the proper order, and it seems like it's probably possible. If a regex can do it, it makes me think that it's probably possible too, and easier, for a tool that "understands" the code like ESLint.

### PR DESCRIPTION
**Pull request by lsdsjy**
Fixes #1463.
This PR reflects the comments from @ota-meshi in #1512 . It yields a fix for each reported error, so eslint-disable comments work normally. Added a test case for eslint-disable comment.

### CODE REVIEW

File: tests/lib/rules/component-tags-order.js [hunk 1]
```diff
@@ -268,7 +345,23 @@ tester.run('component-tags-order', rule, {
...
         }
-      ]
+      ],
+      output: '\n        <script></script>\n        <style></style>\n      '
     }
   ]
 })
+
+describe('suppress reporting with eslint-disable-next-line', () => {
+  it('disable all rules if <!-- eslint-disable-line -->', async () => {
```
Comment (FloEdelmann) [inline tests/lib/rules/component-tags-order.js:355]: The test name should probably be adjusted.

**Review by FloEdelmann — APPROVED** [review]
Thanks for implementing this!

Looks fine from my side. I have just one minor suggestion.

File: tests/lib/rules/component-tags-order.js (hunk 1 above)
Comment (lsdsjy) [reply tests/lib/rules/component-tags-order.js]: Nice catch! It was copied from another test and I forgot to modify the test description.

Comment (lsdsjy) [reply tests/lib/rules/component-tags-order.js]: Updated.

**Review by FloEdelmann — APPROVED** [review]

**Comment by lsdsjy** [comment]
Wait, there's something I can't determine here. If there is a comment right above the element start tag, should we move the comment too? For example,
```vue
<style></style>
<!-- this is a comment -->
<script>
</script>
```

Should we move the comment together with the `<script>` element? What if the comment is a eslint-disable comment, like

```vue
<style></style>
<!-- eslint-disable-next-line vue/multi-word-component-names -->
<script>
export default {
  name: 'Comp'
}
</script>
```

@FloEdelmann @ota-meshi Would you have a look at this plz?

**Comment by FloEdelmann** [comment]
Moving a comment together with the following block would be nice, regardless of whether it is an eslint-disable comment or not. However, I don't think that this would be merge-blocking for me.

File: tests/lib/rules/component-tags-order.js [hunk 2]
```diff
@@ -268,7 +345,23 @@ tester.run('component-tags-order', rule, {
...
+    const code = `<style></style><template></template>
+    <!-- eslint-disable-next-line vue/component-tags-order -->
+    <script></script>`
+    const [{ messages }] = await eslint.lintText(code, { filePath: 'test.vue' })
+    assert.strictEqual(messages.length, 1)
+    // should not fix <script>
+    assert.strictEqual(
+      messages[0].fix.text,
+      '<template></template><style></style>'
+    )
```
Comment (ota-meshi) [inline tests/lib/rules/component-tags-order.js:365]: I think it's better to check the fixed text using the `output` provided by the ESLint class. To use the `output`, add `fix: true` to the ESLint constructor option.

```suggestion
    const [{ messages, output }] = await eslint.lintText(code, { filePath: 'test.vue' })
    assert.strictEqual(messages.length, 0)
    // should not fix <script>
    assert.strictEqual(
      output,
      `<template></template><style></style>
    <!-- eslint-disable-next-line vue/component-tags-order -->
    <script></script>`
    )
```

File: tests/lib/rules/component-tags-order.js [hunk 3]
```diff
@@ -9,6 +9,25 @@
...
+      ecmaVersion: 2015
+    },
+    plugins: ['vue'],
+    rules: {
+      'vue/comment-directive': 'error',
+      'vue/component-tags-order': 'error'
+    }
+  },
+  useEslintrc: false,
+  plugins: { vue: require('../../../lib/index') }
```
Comment (ota-meshi) [inline tests/lib/rules/component-tags-order.js:29]:
```suggestion
  plugins: { vue: require('../../../lib/index') },
  fix: true
```

**Review by ota-meshi — CHANGES_REQUESTED** [review]
Thank you for this PR! I have a change request for testcase.

**Review by ota-meshi — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**